Improve description of "coqc -Q" (#14167)
coq/coq: doc/sphinx/practical-tools/coq-commands.rst

Batch compilation (coqc)
------------------------

The ``coqc`` command compiles a Coq proof script file with a ".v" suffix
to create a compiled file with a ".vo" suffix.  (See :ref:`compiled-files`.)
The components of the filename must be valid Coq identifiers as described in
:ref:`lexical-conventions`; they should contain only letters, digits or
underscores (_) with a ".v" suffix on the final component.
For example ``/bar/foo/toto.v`` is valid, but ``/bar/foo/to-to.v`` is not.

We recommend specifying a logical directory name (which is also the module name)
with the `-R` or the `-Q` options.
Generally we recommend using provided utilities such as `make` to build Coq projects.
See :ref:`building_coq_project`.

.. example:: Compiling and loading a single file

   To compile `foo.v` as part of a module `Mod1` that is rooted
   at `.` (i.e. the directory containing `foo.v`), run `coqc -Q . Mod1 foo.v`.

   To load the module within `coqtop` or `coqide`, use:

      .. coqdoc::

>          Add LoadPath "PATH" as Mod1.

Reviewer: Please don't use this command in example code, it is not recommended to setup your workspace with vernacular commands, but instead using the corresponding flags.

Author: Meaning in _CoqProject or on the command if you must?  At the moment, both require restarting coqtop/coqide.  Perhaps coqide could be modified so it doesn't need to be restarted.

I agree `Add LoadPath` is not great, but it doesn't require restarting CoqIDE, which is a bit disruptive.  So maybe reasonable for quick experiments.

Reviewer 2: For coqide putting the flags in coqproject then closing/reopening the .v file should work.

Author: Less bad, thanks, but it should be more seamless in CoqIDE.  Opening a new scratch buffer doesn't reload the info, which I expected.

Author: Looks like CoqIDE lets you set the name of the project file in preferences.  `coqtop` doesn't seem to apply the default name, nor do I see a command line option for providing the name.  That's a bug, too, even if normal users won't use coqtop.

Reviewer 2: >That's a bug, too, even if normal users won't use coqtop.

What does that mean? coqtop is kept for proffgeneral and that's about it.

Author: We use it in the development process and tests, too.

Reviewer 3: `coqtop` and `coqc` only have `-R` / `-Q` options that you must pass manually. The idea is that users are never supposed to call these commands directly. A user should use an editor, which should load `_CoqProject` and `coq_makefile` or `dune` to build.

Batch compilation (coqc)
------------------------

The ``coqc`` command compiles a Coq proof script file with a ".v" suffix
to create a compiled file with a ".vo" suffix.  (See :ref:`compiled-files`.)
The components of the filename must be valid Coq identifiers as described in
:ref:`lexical-conventions`; they should contain only letters, digits or
underscores (_) with a ".v" suffix on the final component.

> For example ``/bar/foo/toto.v`` is valid, but ``/bar/foo/to-to.v`` is not.

Reviewer 3: There is no restriction on the validity of the directory which contains a `.v` file:

```
$ coqc test\ with\ space/test.v 
```
works.

Author: This is not new text.  How does such a name get mapped to a module name?  I see such names as something to avoid even if they're permitted.

Reviewer 3: You extended the existing text. The previous test only talked about the filename (i.e. last component). I don't know exactly how things work, so I cannot answer your question precisely.

> We recommend specifying a logical directory name (which is also the module name)
> with the `-R` or the `-Q` options.
> Generally we recommend using provided utilities such as `make` to build Coq projects.
> See :ref:`building_coq_project`.

Reviewer 3: :+1:

```suggestion
We recommend specifying a logical directory name (which is also the module name)
with the `-R` or the `-Q` options.
Generally we recommend using provided utilities such as `coq_makefile` or `dune` to build Coq projects.
See :ref:`building_coq_project`.
```

Author: `coq_makefile` creates a makefile that you then run with `make`, right?

Made it: "Generally we recommend using utilities such as `make` or `dune` to build Coq projects,
using `coq_makefile` to generate the dependencies."

Reviewer 3: > `coq_makefile` creates a makefile that you then run with `make`, right?

Yes.
 
> Made it: "Generally we recommend using utilities such as `make` or `dune` to build Coq projects,
> using `coq_makefile` to generate the dependencies."

Maybe tweak it further to:

"Generally we recommend using utilities such as `make` (using `coq_makefile` to generate the `Makefile`) or `dune` to build Coq projects"
> ."

.. example:: Compiling and loading a single file

   If `foo.v` is in Coq's current directory, you can use `coqc foo.v`
   to compile it and then `Require foo.` in your script.  But this
   doesn't scale well for larger projects.

   Generally it's' better to define a new module:
   To compile `foo.v` as part of a module `Mod1` that is rooted
   at `.` (i.e. the directory containing `foo.v`), run `coqc -Q . Mod1 foo.v`.

   To make the module available in `CoqIDE`, include the following lines in the
   `_CoqProject` file (see :ref:`coq_makefile`) in the directory from which you
   start `CoqIDE`.  For now, you need to close and reload a named
   script file for `CoqIDE` to pick up the change or restart `CoqIDE`.
   The project file name is configurable in `Edit / Preferences / Project`.

      .. coqdoc::

>          -R ../empty_dir/mod1 Mod1

Reviewer 3: What is `empty_dir` supposed to represent? Wouldn't it be better to just recommend that the `_CoqProject` file be always located at the root of a project and that CoqIDE (or another compatible editor) be always run from the root of the project as well?

Author: I removed empty_dir, that was my local test.

Having to start CoqIDE in a specific directory is inconvenient if you start it from an icon on your desktop.  Icons don't allow for changing the directory.  CoqIDE should let you pick the project within its GUI.  I think there are probably some small code tweaks we could make that would make for a better user experience.

Reviewer 3: I agree.

>          ./foo.v

Reviewer 3: Note that listing files is only useful for `coq_makefile`. Only the options on the first lines matter for editors like CoqIDE.

Author: I replaced with a vertical ellipsis (a colon).

Reviewer 3: Is this not going to be interpreted as a literal colon by some users?

>          -R ../mod1 Mod1

Reviewer 3: It's still not clear to me why you use `../mod1` here instead of `.` as in:
```suggestion
         -R . Mod1
```

Author: I did something better than your suggestion.

Updated.

   It's also possible to load a module within `coqtop` or `coqide` with
   these commands like this.  The drawback of this is that it adds
   environment-specific information (the PATH) to your script, making

>    it non-portable.

Reviewer 3: ```suggestion
   it non-portable. Therefore, this is discouraged.
```

Reviewer: It could also be the case that these commands get removed in the future, as they commit to a library organization model which may not hold. Personally, I wouldn't recommend them under any circumstance.

Author: We document Coq as it is, not as we want it to become.  Only by mentioning and explaining this feature can we say not to use it.  Feel free to start a discussion on removing it--though I think we should review how to use the various compilation options first and perhaps tweak them.
